Socketmode panic, write on closed channel

### What happened
```
panic: send on closed channel

goroutine 47 [running]:
github.com/slack-go/slack/socketmode.(*Client).receiveMessagesInto(0x14000243950, {0x101c6e948, 0x1400043cac0}, 0x14000676c60, 0x1400061a6c0)
        /Users/censored/go/pkg/mod/github.com/slack-go/slack@v0.11.3/socketmode/socket_mode_managed_conn.go:484 +0x64c
github.com/slack-go/slack/socketmode.(*Client).runMessageReceiver(0x14000243950, {0x101c6e948, 0x1400043cac0}, 0x14000676c60, 0x1400061a6c0)
        /Users/censored/go/pkg/mod/github.com/slack-go/slack@v0.11.3/socketmode/socket_mode_managed_conn.go:363 +0x50
github.com/slack-go/slack/socketmode.(*Client).run.func4()
        /Users/censored/go/pkg/mod/github.com/slack-go/slack@v0.11.3/socketmode/socket_mode_managed_conn.go:133 +0x94
created by github.com/slack-go/slack/socketmode.(*Client).run
        /Users/censored/go/pkg/mod/github.com/slack-go/slack@v0.11.3/socketmode/socket_mode_managed_conn.go:128 +0x948
Exiting.
```

Looks like channel got closed:
https://github.com/slack-go/slack/blob/6f5eda2a6c1e50c8430bb07657bbe8be0def2b37/socketmode/socket_mode_managed_conn.go#L58
before message got passed to the channel:
https://github.com/slack-go/slack/blob/6f5eda2a6c1e50c8430bb07657bbe8be0def2b37/socketmode/socket_mode_managed_conn.go#L485
because we don't wait for completion:
https://github.com/slack-go/slack/blob/6f5eda2a6c1e50c8430bb07657bbe8be0def2b37/socketmode/socket_mode_managed_conn.go#L126

### Expected behavior
No panic

### Steps to reproduce
Start socketmode slack client

#### reproducible code
https://github.com/slack-go/slack/blob/master/examples/socketmode/socketmode.go

### Versions
- Go: 1.18
- slack-go/slack: 0.11.3

--- a/socketmode/socket_mode_managed_conn.go
+++ b/socketmode/socket_mode_managed_conn.go
@@ -123,9 +123,10 @@ func (smc *Client) run(ctx context.Context, connectionCount int) error {
 		}
 	}()
 
-	// We don't wait on runMessageReceiver because it doesn't block on a select with the context,
-	// so we'd have to wait for the ReadJSON to time out, which can take a while.
+	// Need to wait for runMessageReceiver to avoid panic described in https://github.com/slack-go/slack/issues/1125
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
 		defer cancel()
 
 		// The receiver reads WebSocket messages, and enqueues parsed Socket Mode requests to be handled by
@@ -439,8 +440,40 @@ func (smc *Client) receiveMessagesInto(ctx context.Context, conn *websocket.Conn
 	smc.Debugf("Starting to receive message")
 	defer smc.Debugf("Finished to receive message")
 
+	var err error
+
 	event := json.RawMessage{}
-	err := conn.ReadJSON(&event)
+
+	readJsonErr := make(chan error)
+	go func() {
+		select {
+		case readJsonErr <- conn.ReadJSON(&event):
+			// if conn.ReadJSON method returns after ctx.Done(), no one will read the unbuffered channel
+			// so, to avoid goroutines leak we need to check for ctx.Done() here too.
+			// need to say here, that conn.ReadJSON will really return after ctx.Done(), because in that case
+			// conn is closed
+			break
+		case <-ctx.Done():
+			// just need to listen ctx.Done, nothing has to be done here
+			break
+		}
+	}()
+
+	select {
+	case err = <-readJsonErr:
+		// we have awaited response from conn.ReadJSON, so, we can handle error now
+		break
+	case <-ctx.Done():
+		// context cancellation signal got, closing connection and returning
+		cerr := conn.Close()
+		if cerr != nil {
+			smc.Debugf("Failed to close connection: %v", cerr)
+		}
+
+		return ctx.Err()
+	}
+
+	// after select above, the err will be set to result from conn.ReadJSON, handling
 
 	// check if the connection was closed.
 	if websocket.IsUnexpectedCloseError(err) {

